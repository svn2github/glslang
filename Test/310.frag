#version 310 es
highp float nodef3(float); // ERROR, no default precision
precision mediump float;
precision highp usampler2D;
precision highp sampler2D;
precision highp isampler2DArray;

layout(origin_upper_left, pixel_center_integer) in vec4 gl_FragCoord;  // ERROR, not supported

layout(location = 2) in vec3 v3;
layout(location = 2) in mat4 yi;  // ERROR, locations conflict with xi

uniform sampler2D arrayedSampler[5];
uniform usampler2D usamp2d;
uniform usampler2DRect samp2dr;      // ERROR, reserved
uniform isampler2DArray isamp2DA;

in vec2 c2D;
uniform int i;

void main()
{
    vec4 v = texture(arrayedSampler[i], c2D);  // ERROR

    ivec2 offsets[4];
    const ivec2 constOffsets[4] = ivec2[4](ivec2(1,2), ivec2(3,4), ivec2(15,16), ivec2(-2,0));
    uvec4 uv4 = textureGatherOffsets(samp2dr, c2D, offsets, 2);  // ERROR, not supported
    vec4 v4 = textureGather(arrayedSampler[0], c2D);
    ivec4 iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 3);
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), i);  // ERROR, last argument not const
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 4);  // ERROR, last argument out of range
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 1+2);
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(i));
}

out vec4 outp;

void foo23()
{
    const ivec2[3] offsets = ivec2[3](ivec2(1,2), ivec2(3,4), ivec2(15,16));

    textureProjGradOffset(usamp2d, outp, vec2(0.0), vec2(0.0), ivec2(c2D));     // ERROR, offset not constant
    textureProjGradOffset(usamp2d, outp, vec2(0.0), vec2(0.0), offsets[1]);
    textureProjGradOffset(usamp2d, outp, vec2(0.0), vec2(0.0), offsets[2]);     // ERROR, offset out of range
    textureProjGradOffset(usamp2d, outp, vec2(0.0), vec2(0.0), ivec2(-10, 20)); // ERROR, offset out of range

    if (gl_HelperInvocation)
        ++outp;

    int sum = gl_MaxVertexImageUniforms +
              gl_MaxFragmentImageUniforms +
              gl_MaxComputeImageUniforms +
              gl_MaxCombinedImageUniforms +
              gl_MaxCombinedShaderOutputResources;

    bool b1, b2, b3, b;

    b1 = mix(b2, b3, b);
    uvec3 um3 = mix(uvec3(i), uvec3(i), bvec3(b));
    ivec4 im4 = mix(ivec4(i), ivec4(i), bvec4(b));
}

layout(binding=3) uniform sampler2D s1;
layout(binding=3) uniform sampler2D s2; // ERROR: overlapping bindings?  Don't see that in the 310 spec.
layout(binding=2) uniform writeonly image2D      i2D;
layout(binding=4) uniform readonly  image3D      i3D;
layout(binding=5) uniform           imageCube    iCube;
layout(binding=6) uniform           image2DArray i2DA;
layout(binding=6) uniform coherent volatile restrict image2D i2Dqualified;

layout(binding = 1) uniform bb {
    int foo;
    layout(binding = 2) float f;     // ERROR
} bbi;

in centroid vec4 centroidIn;
layout(location = 200000) uniform vec4 bigl;  // ERROR, location too big

layout(early_fragment_tests) in;

layout(location = 40) out vec4 bigout1;  // ERROR, too big
layout(location = 40) out vec4 bigout2;  // ERROR, overlap
layout(location = -2) out vec4 neg;      // ERROR, negative

layout(std430) buffer b430 {
    int i;
} b430i;

layout(shared) uniform bshar {
    int i;
} bshari;

in smooth vec4 smoothIn;
in flat int flatIn;

uniform sampler2DMS s2dms;

void foots()
{
    highp ivec2 v2 = textureSize(s1, 2);
    highp ivec3 v3 = textureSize(isamp2DA, 3);
    v2 = textureSize(s2dms);
    v2 = imageSize(i2D);
    v3 = imageSize(i3D);
    v2 = imageSize(iCube);
    v3 = imageSize(i2DA);
    v2 = imageSize(i2Dqualified);
}

out bool bout;          // ERROR
out image2D imageOut;   // ERROR
out mat2x3 mout;        // ERROR

in bool inb;         // ERROR
in sampler2D ino;    // ERROR
in float ina[4];
in float inaa[4][2]; // ERROR
struct S { float f; };
in S ins;
in S[4] inasa;       // ERROR
in S insa[4];        // ERROR
struct SA { float f[4]; };
in SA inSA;          // ERROR
struct SS { float f; S s; };
in SS inSS;          // ERROR

#ifndef GL_EXT_shader_io_blocks
#error GL_EXT_shader_io_blocks not defined
#endif

#extension GL_EXT_shader_io_blocks : enable

out outbname { int a; } outbinst;   // ERROR, not out block in fragment shader

in inbname {
    int a;
    vec4 v;
    struct { int b; } s;     // ERROR, nested struct definition
} inbinst;

in inbname2 {
    layout(location = 12) int aAnon;
    layout(location = 13) centroid in vec4 vAnon;
};

in layout(location = 13) vec4 aliased; // ERROR, aliased

in inbname2 {                // ERROR, reuse of block name
    int aAnon;
    centroid in vec4 vAnon;
};

in badmember {               // ERROR, aAnon already in global scope
    int aAnon;
};

int inbname;                 // ERROR, redefinition of block name

vec4 vAnon;                  // ERROR, anon in global scope; redefinition

in arrayed {
    float f;
} arrayedInst[4];

void fooIO()
{
    vec4 v = inbinst.v + vAnon;
    v *= arrayedInst[2].f;
    v *= arrayedInst[i].f;       // ERROR, not constant
}

in vec4 gl_FragCoord;
layout(origin_upper_left, pixel_center_integer) in vec4 gl_FragCoord;  // ERROR, non-ES

layout(early_fragment_tests) in;
out float gl_FragDepth;
layout(depth_any) out float gl_FragDepth;  // ERROR, non-ES

void foo_IO()
{
    gl_FragDepth = 0.2;  // ERROR, early_fragment_tests declared
}

out float gl_FragDepth;
