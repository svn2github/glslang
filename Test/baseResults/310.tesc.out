310.tesc
Warning, version 310 is not yet complete; most version-specific features are present, but some are missing.
ERROR: 0:8: 'quads' : unrecognized layout identifier, or qualifier requires assignment (e.g., binding = 4) 
ERROR: 0:9: 'ccw' : unrecognized layout identifier, or qualifier requires assignment (e.g., binding = 4) 
ERROR: 0:10: 'fractional_even_spacing' : unrecognized layout identifier, or qualifier requires assignment (e.g., binding = 4) 
ERROR: 0:12: 'patch' : can only use on output in tessellation-control shader 
ERROR: 0:27: 'gl_ClipDistance' : no such field in structure 
ERROR: 0:27: 'expression' :  left of '[' is not of type array, matrix, or vector  
ERROR: 0:35: 'gl_ClipDistance' : no such field in structure 
ERROR: 0:35: 'expression' :  left of '[' is not of type array, matrix, or vector  
ERROR: 0:35: 'assign' :  l-value required (can't modify a const)
ERROR: 0:41: '' : tessellation control barrier() cannot be placed within flow control 
ERROR: 0:43: '' : tessellation control barrier() cannot be placed within flow control 
ERROR: 0:48: '' : tessellation control barrier() cannot be placed within flow control 
ERROR: 0:53: '' : tessellation control barrier() cannot be placed within flow control 
ERROR: 0:56: '' : tessellation control barrier() cannot be placed within flow control 
ERROR: 0:63: '' : tessellation control barrier() cannot be placed after a return from main() 
ERROR: 0:66: 'vertices' : can only apply to 'out' 
ERROR: 0:67: 'vertices' : cannot change previously set layout value 
ERROR: 0:71: '[' :  array index out of range '4'
ERROR: 0:73: '' : tessellation control barrier() must be in main() 
ERROR: 0:76: 'in' : type must be an array: ina
ERROR: 0:78: '[]' : tessellation input array size must be gl_MaxPatchVertices or implicitly sized 
ERROR: 0:80: '' : array size required 
ERROR: 0:86: 'location' : overlapping use of location 4
ERROR: 0:90: 'location' : overlapping use of location 4
ERROR: 0:94: 'precise' : Reserved word. 
ERROR: 0:95: 'fma' : no matching overloaded function found 
ERROR: 0:104: 'sample' : Reserved word. 
ERROR: 0:106: 'vertices' : can only apply to a standalone qualifier 
ERROR: 0:107: 'vertices' : inconsistent output number of vertices for array size of misSized
ERROR: 29 compilation errors.  No code generated.


Shader version: 310
Requested GL_ARB_separate_shader_objects
Requested GL_OES_shader_io_blocks
Requested GL_OES_tessellation_shader
vertices = 4
ERROR: node is still EOpNull!
0:15  Function Definition: main( (global void)
0:15    Function Parameters: 
0:17    Sequence
0:17      Barrier (global void)
0:19      Sequence
0:19        move second child to first child (temp highp int)
0:19          'a' (temp highp int)
0:19          Constant:
0:19            5392 (const int)
0:25      Sequence
0:25        move second child to first child (temp highp 4-component vector of float)
0:25          'p' (temp highp 4-component vector of float)
0:25          gl_Position: direct index for structure (in highp 4-component vector of float)
0:25            direct index (temp block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:25              'gl_in' (in 32-element array of block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:25              Constant:
0:25                1 (const int)
0:25            Constant:
0:25              0 (const int)
0:26      Sequence
0:26        move second child to first child (temp highp float)
0:26          'ps' (temp highp float)
0:26          gl_PointSize: direct index for structure (in highp float)
0:26            direct index (temp block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:26              'gl_in' (in 32-element array of block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:26              Constant:
0:26                1 (const int)
0:26            Constant:
0:26              1 (const int)
0:27      Sequence
0:27        move second child to first child (temp highp float)
0:27          'cd' (temp highp float)
0:27          Constant:
0:27            0.000000
0:29      Sequence
0:29        move second child to first child (temp highp int)
0:29          'pvi' (temp highp int)
0:29          'gl_PatchVerticesIn' (in highp int PatchVertices)
0:30      Sequence
0:30        move second child to first child (temp highp int)
0:30          'pid' (temp highp int)
0:30          'gl_PrimitiveID' (in highp int PrimitiveID)
0:31      Sequence
0:31        move second child to first child (temp highp int)
0:31          'iid' (temp highp int)
0:31          'gl_InvocationID' (in highp int InvocationID)
0:33      move second child to first child (temp highp 4-component vector of float)
0:33        gl_Position: direct index for structure (out highp 4-component vector of float Position)
0:33          direct index (temp block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:33            'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:33            Constant:
0:33              1 (const int)
0:33          Constant:
0:33            0 (const int)
0:33        'p' (temp highp 4-component vector of float)
0:34      move second child to first child (temp highp float)
0:34        gl_PointSize: direct index for structure (out highp float PointSize)
0:34          direct index (temp block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:34            'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:34            Constant:
0:34              1 (const int)
0:34          Constant:
0:34            1 (const int)
0:34        'ps' (temp highp float)
0:35      move second child to first child (temp highp float)
0:35        Constant:
0:35          0.000000
0:35        'cd' (temp highp float)
0:37      move second child to first child (temp highp float)
0:37        direct index (patch temp highp float TessLevelOuter)
0:37          'gl_TessLevelOuter' (patch out 4-element array of highp float TessLevelOuter)
0:37          Constant:
0:37            3 (const int)
0:37        Constant:
0:37          3.200000
0:38      move second child to first child (temp highp float)
0:38        direct index (patch temp highp float TessLevelInner)
0:38          'gl_TessLevelInner' (patch out 2-element array of highp float TessLevelInner)
0:38          Constant:
0:38            1 (const int)
0:38        Constant:
0:38          1.300000
0:40      Test condition and select (temp void)
0:40        Condition
0:40        Compare Greater Than (temp bool)
0:40          'a' (temp highp int)
0:40          Constant:
0:40            10 (const int)
0:40        true case
0:41        Barrier (global void)
0:40        false case
0:43        Barrier (global void)
0:45      Barrier (global void)
0:49      Loop with condition not tested first
0:49        Loop Condition
0:49        Compare Greater Than (temp bool)
0:49          'a' (temp highp int)
0:49          Constant:
0:49            10 (const int)
0:49        Loop Body
0:48        Sequence
0:48          Barrier (global void)
0:51      switch
0:51      condition
0:51        'a' (temp highp int)
0:51      body
0:51        Sequence
0:52          default: 
0:?           Sequence
0:53            Barrier (global void)
0:54            Branch: Break
0:56      Test condition and select (temp highp int)
0:56        Condition
0:56        Compare Less Than (temp bool)
0:56          'a' (temp highp int)
0:56          Constant:
0:56            12 (const int)
0:56        true case
0:56        'a' (temp highp int)
0:56        false case
0:56        Comma (temp highp int)
0:56          Barrier (global void)
0:56          'a' (temp highp int)
0:58      Sequence
0:58        Barrier (global void)
0:61      Branch: Return
0:63      Barrier (global void)
0:69  Function Definition: foo( (global void)
0:69    Function Parameters: 
0:71    Sequence
0:71      gl_Position: direct index for structure (out highp 4-component vector of float Position)
0:71        direct index (temp block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:71          'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:71          Constant:
0:71            4 (const int)
0:71        Constant:
0:71          0 (const int)
0:73      Barrier (global void)
0:92  Function Definition: foop( (global void)
0:92    Function Parameters: 
0:?     Sequence
0:95      move second child to first child (temp highp float)
0:95        'd' (temp highp float)
0:95        Constant:
0:95          0.000000
0:?   Linker Objects
0:?     'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:?     'outa' (global 4-element array of highp int)
0:?     'patchIn' (patch in highp 4-component vector of float)
0:?     'patchOut' (patch out highp 4-component vector of float)
0:?     'ina' (in highp 2-component vector of float)
0:?     'inb' (in 32-element array of highp 2-component vector of float)
0:?     'inc' (in 32-element array of highp 2-component vector of float)
0:?     'ind' (in 32-element array of highp 2-component vector of float)
0:?     'implA' (patch out implicitly-sized array of highp float)
0:?     'ivla' (layout(location=3 ) in 32-element array of highp 4-component vector of float)
0:?     'ivlb' (layout(location=4 ) in 32-element array of highp 4-component vector of float)
0:?     'ivlc' (layout(location=4 ) in 32-element array of highp 4-component vector of float)
0:?     'ovla' (layout(location=3 ) out 4-element array of highp 4-component vector of float)
0:?     'ovlb' (layout(location=4 ) out 4-element array of highp 4-component vector of float)
0:?     'ovlc' (layout(location=4 ) out 4-element array of highp 4-component vector of float)
0:?     'pinbi' (patch out block{out highp int a})
0:?     'myColor2' (centroid out 4-element array of highp 3-component vector of float)
0:?     'centr' (centroid in 32-element array of highp 3-component vector of float)
0:?     'perSampleColor' (sample out 4-element array of highp 4-component vector of float)
0:?     'badlay' (out 4-element array of highp float)
0:?     'misSized' (out 5-element array of highp float)
0:?     'okaySize' (out 4-element array of highp float)


Linked tessellation control stage:


Shader version: 310
Requested GL_ARB_separate_shader_objects
Requested GL_OES_shader_io_blocks
Requested GL_OES_tessellation_shader
vertices = 4
ERROR: node is still EOpNull!
0:15  Function Definition: main( (global void)
0:15    Function Parameters: 
0:17    Sequence
0:17      Barrier (global void)
0:19      Sequence
0:19        move second child to first child (temp highp int)
0:19          'a' (temp highp int)
0:19          Constant:
0:19            5392 (const int)
0:25      Sequence
0:25        move second child to first child (temp highp 4-component vector of float)
0:25          'p' (temp highp 4-component vector of float)
0:25          gl_Position: direct index for structure (in highp 4-component vector of float)
0:25            direct index (temp block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:25              'gl_in' (in 32-element array of block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:25              Constant:
0:25                1 (const int)
0:25            Constant:
0:25              0 (const int)
0:26      Sequence
0:26        move second child to first child (temp highp float)
0:26          'ps' (temp highp float)
0:26          gl_PointSize: direct index for structure (in highp float)
0:26            direct index (temp block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:26              'gl_in' (in 32-element array of block{in highp 4-component vector of float gl_Position, in highp float gl_PointSize})
0:26              Constant:
0:26                1 (const int)
0:26            Constant:
0:26              1 (const int)
0:27      Sequence
0:27        move second child to first child (temp highp float)
0:27          'cd' (temp highp float)
0:27          Constant:
0:27            0.000000
0:29      Sequence
0:29        move second child to first child (temp highp int)
0:29          'pvi' (temp highp int)
0:29          'gl_PatchVerticesIn' (in highp int PatchVertices)
0:30      Sequence
0:30        move second child to first child (temp highp int)
0:30          'pid' (temp highp int)
0:30          'gl_PrimitiveID' (in highp int PrimitiveID)
0:31      Sequence
0:31        move second child to first child (temp highp int)
0:31          'iid' (temp highp int)
0:31          'gl_InvocationID' (in highp int InvocationID)
0:33      move second child to first child (temp highp 4-component vector of float)
0:33        gl_Position: direct index for structure (out highp 4-component vector of float Position)
0:33          direct index (temp block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:33            'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:33            Constant:
0:33              1 (const int)
0:33          Constant:
0:33            0 (const int)
0:33        'p' (temp highp 4-component vector of float)
0:34      move second child to first child (temp highp float)
0:34        gl_PointSize: direct index for structure (out highp float PointSize)
0:34          direct index (temp block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:34            'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:34            Constant:
0:34              1 (const int)
0:34          Constant:
0:34            1 (const int)
0:34        'ps' (temp highp float)
0:35      move second child to first child (temp highp float)
0:35        Constant:
0:35          0.000000
0:35        'cd' (temp highp float)
0:37      move second child to first child (temp highp float)
0:37        direct index (patch temp highp float TessLevelOuter)
0:37          'gl_TessLevelOuter' (patch out 4-element array of highp float TessLevelOuter)
0:37          Constant:
0:37            3 (const int)
0:37        Constant:
0:37          3.200000
0:38      move second child to first child (temp highp float)
0:38        direct index (patch temp highp float TessLevelInner)
0:38          'gl_TessLevelInner' (patch out 2-element array of highp float TessLevelInner)
0:38          Constant:
0:38            1 (const int)
0:38        Constant:
0:38          1.300000
0:40      Test condition and select (temp void)
0:40        Condition
0:40        Compare Greater Than (temp bool)
0:40          'a' (temp highp int)
0:40          Constant:
0:40            10 (const int)
0:40        true case
0:41        Barrier (global void)
0:40        false case
0:43        Barrier (global void)
0:45      Barrier (global void)
0:49      Loop with condition not tested first
0:49        Loop Condition
0:49        Compare Greater Than (temp bool)
0:49          'a' (temp highp int)
0:49          Constant:
0:49            10 (const int)
0:49        Loop Body
0:48        Sequence
0:48          Barrier (global void)
0:51      switch
0:51      condition
0:51        'a' (temp highp int)
0:51      body
0:51        Sequence
0:52          default: 
0:?           Sequence
0:53            Barrier (global void)
0:54            Branch: Break
0:56      Test condition and select (temp highp int)
0:56        Condition
0:56        Compare Less Than (temp bool)
0:56          'a' (temp highp int)
0:56          Constant:
0:56            12 (const int)
0:56        true case
0:56        'a' (temp highp int)
0:56        false case
0:56        Comma (temp highp int)
0:56          Barrier (global void)
0:56          'a' (temp highp int)
0:58      Sequence
0:58        Barrier (global void)
0:61      Branch: Return
0:63      Barrier (global void)
0:69  Function Definition: foo( (global void)
0:69    Function Parameters: 
0:71    Sequence
0:71      gl_Position: direct index for structure (out highp 4-component vector of float Position)
0:71        direct index (temp block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:71          'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:71          Constant:
0:71            4 (const int)
0:71        Constant:
0:71          0 (const int)
0:73      Barrier (global void)
0:92  Function Definition: foop( (global void)
0:92    Function Parameters: 
0:?     Sequence
0:95      move second child to first child (temp highp float)
0:95        'd' (temp highp float)
0:95        Constant:
0:95          0.000000
0:?   Linker Objects
0:?     'gl_out' (out 4-element array of block{out highp 4-component vector of float Position gl_Position, out highp float PointSize gl_PointSize})
0:?     'outa' (global 4-element array of highp int)
0:?     'patchIn' (patch in highp 4-component vector of float)
0:?     'patchOut' (patch out highp 4-component vector of float)
0:?     'ina' (in highp 2-component vector of float)
0:?     'inb' (in 32-element array of highp 2-component vector of float)
0:?     'inc' (in 32-element array of highp 2-component vector of float)
0:?     'ind' (in 32-element array of highp 2-component vector of float)
0:?     'implA' (patch out 1-element array of highp float)
0:?     'ivla' (layout(location=3 ) in 32-element array of highp 4-component vector of float)
0:?     'ivlb' (layout(location=4 ) in 32-element array of highp 4-component vector of float)
0:?     'ivlc' (layout(location=4 ) in 32-element array of highp 4-component vector of float)
0:?     'ovla' (layout(location=3 ) out 4-element array of highp 4-component vector of float)
0:?     'ovlb' (layout(location=4 ) out 4-element array of highp 4-component vector of float)
0:?     'ovlc' (layout(location=4 ) out 4-element array of highp 4-component vector of float)
0:?     'pinbi' (patch out block{out highp int a})
0:?     'myColor2' (centroid out 4-element array of highp 3-component vector of float)
0:?     'centr' (centroid in 32-element array of highp 3-component vector of float)
0:?     'perSampleColor' (sample out 4-element array of highp 4-component vector of float)
0:?     'badlay' (out 4-element array of highp float)
0:?     'misSized' (out 5-element array of highp float)
0:?     'okaySize' (out 4-element array of highp float)

